Enable HorizontalPodAutoscaler Support for DeploymentConfigs

DirectXMan12 opened the pull request:
This PR adds support for HPAs scaling DeploymentConfigs.   A new "HybridClient" struct is introduced to support Namespacers that need to point at resources or subresources that could be in either the Origin API or the Kube API (like `Scale`).

This commit also imports the Scale kind into the Origin v1 API (but it's still "backed" in the unversioned API by the underlying Kube resource).

Finally, this commit contains a patch to Kube that converts the HPA controller from using a single `Client` type to using separate `Namespacers` for its various needs.


--- 
TODOs
 - [x] cherry-pick kubernetes/kubernetes#16570 and update policy.
 - [x] cherry pick kubernetes/kubernetes#15706 from upstream into this PR (underneath your changes
 - [x] open issue for " fix the metrics client to take pod names, rather than requiring cluster-wide pod-listing privileges" https://github.com/openshift/origin/pull/5310/files#r43427148 (DONE: kubernetes/kubernetes#16675)
 - [x] rename UPSTREAM commits to match `UPSTREAM: PR` or `UPSTREAM: <carry>` format
 - [x] remove WIP from commit name
 - [ ] open issue for " fix the client to no longer need root proxy access" https://github.com/openshift/origin/pull/5310/files#r43427078
 - [ ] add end-to-end or extended test in origin to prove that this works without default (and comparatively restrictive) policy.
 - [x] add a comment to `Update` to document the order in which the deploymentconfig controller does the updates to resources as it creates new RCs and changes its RC's replicas.  That information is required to reason about race potential.
 - [x] resolve issue with cross-namespace HPA requests (prevent it and add a test ensuring it is prevented) https://github.com/openshift/origin/pull/5579

DirectXMan12 commented:
@ncdc @deads2k

DirectXMan12 commented:
Once #5286 lands, I'll rebase on top of that.

@ncdc @smarterclayton this PR currently follows the upstream Kube in that it updates the  replicas field on the actual deployment controller.  It occurs to me that this does not appear to do much in Origin (since future deployments seem to be based on the Scale of the previous deployment).  Should I update so that it affects the latest deployment itself (like the scale command in Origin)?

ncdc commented:
@ironcladlou

ncdc commented:
@DirectXMan12 please use separate commits for Godeps modifications, and make sure the commit message starts with UPSTREAM. Thanks!

smarterclayton commented:
Answer is yes.

On Wed, Oct 21, 2015 at 8:40 PM, Andy Goldstein <notifications@github.com>
wrote:

> @DirectXMan12 <https://github.com/DirectXMan12> please use separate
> commits for Godeps modifications, and make sure the commit message starts
> with UPSTREAM. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5310#issuecomment-150064012>.
>

liggitt commented on pkg/client/client.go:
```diff
@@ -297,3 +297,16 @@ func DefaultOpenShiftUserAgent() string {
 	return fmt.Sprintf("%s/%s (%s/%s) openshift/%s", path.Base(os.Args[0]), version, runtime.GOOS, runtime.GOARCH, commit)
 }
+
+type HybridClient struct {
```
add `var _ = unversioned.ScaleNamespacer(&{HybridClient})` to clarify and enforce the intent

deads2k replied:
Nix this.  Its been rejected on principle mulitple times.

ncdc replied:
Nix what?

liggitt replied:
@deads2k joint client?

deads2k replied:
> @deads2k joint client?

Yes.  `HybridClient` should not exist.  You should specify what you want in your API.

ncdc replied:
We need both.

kargakis replied:
Have a look at our dc scaler and reaper implementations. They use both of the client under the hood w/o exposing a hybrid client.

ncdc replied:
This may not be the right way to implement it, but we have to have a type that implements the kube client ScaleInterface that supports both origin and kube types.

kargakis replied:
Ok, I think that our [Client](https://github.com/openshift/origin/blob/43d027cdcccfbbd10b48ea498ca8e0b9c51cb63c/pkg/client/client.go#L238-L241) should include the `ExtensionsClient` and then internally mix the clients like we already do for `scale` and `delete`.

kargakis replied:
See how the Kubernetes client is currently implemented: https://github.com/kubernetes/kubernetes/blob/071d21257fdfd439a9286fffbd5972c85643cd49/pkg/client/unversioned/client.go#L128-l134

We should do the same for our client

deads2k replied:
> See how the Kubernetes client is currently implemented: https://github.com/kubernetes/kubernetes/blob/071d21257fdfd439a9286fffbd5972c85643cd49/pkg/client/unversioned/client.go#L128-l134
> 
> We should do the same for our client

This is what they do
```go
// Client is the implementation of a Kubernetes client.
type Client struct {
	*RESTClient
	*ExtensionsClient
	// TODO: remove this when we re-structure pkg/client.
	*DiscoveryClient
}
```

I don't think that this is a good thing.  Having a single interface like that encourages non-specificity and it does not scale well as you add different clients.  I think that a reasonable division is on the APIGroup boundary.  Using different clients for different APIGroups will help keep the dependency structure clear, prevents a mega-client problem, and allows a clear expectation as APIGroups are enabled and disabled.

@smarterclayton as I recall, you were not a fan combined `Clients`.

smarterclayton replied:
I am not, we should be providing an implementation of ScaleNamespacer and
ScaleInterface that multiplexes to clients based on known kinds.

On Thu, Oct 22, 2015 at 1:35 PM, David Eads <notifications@github.com>
wrote:

> In pkg/client/client.go
> <https://github.com/openshift/origin/pull/5310#discussion_r42779351>:
>
> > @@ -297,3 +297,16 @@ func DefaultOpenShiftUserAgent() string {
> >  	version = seg[0]
> >  	return fmt.Sprintf("%s/%s (%s/%s) openshift/%s", path.Base(os.Args[0]), version, runtime.GOOS, runtime.GOARCH, commit)
> >  }
> > +
> > +type HybridClient struct {
>
> See how the Kubernetes client is currently implemented:
> https://github.com/kubernetes/kubernetes/blob/071d21257fdfd439a9286fffbd5972c85643cd49/pkg/client/unversioned/client.go#L128-l134
>
> We should do the same for our client
>
> This is what they do
>
> // Client is the implementation of a Kubernetes client.type Client struct {
>     *RESTClient
>     *ExtensionsClient
>     // TODO: remove this when we re-structure pkg/client.
>     *DiscoveryClient
> }
>
> I don't think that this is a good thing. Having a single interface like
> that encourages non-specificity and it does not scale well as you add
> different clients. I think that a reasonable division is on the APIGroup
> boundary. Using different clients for different APIGroups will help keep
> the dependency structure clear, prevents a mega-client problem, and allows
> a clear expectation as APIGroups are enabled and disabled.
>
> @smarterclayton <https://github.com/smarterclayton> as I recall, you were
> not a fan combined Clients.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5310/files#r42779351>.
>

smarterclayton replied:
Or even better, uses "IsOriginResource" to select the origin kinds, and
defaults all other to kube.

On Thu, Oct 22, 2015 at 2:13 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I am not, we should be providing an implementation of ScaleNamespacer and
> ScaleInterface that multiplexes to clients based on known kinds.
>
> On Thu, Oct 22, 2015 at 1:35 PM, David Eads <notifications@github.com>
> wrote:
>
>> In pkg/client/client.go
>> <https://github.com/openshift/origin/pull/5310#discussion_r42779351>:
>>
>> > @@ -297,3 +297,16 @@ func DefaultOpenShiftUserAgent() string {
>> >  	version = seg[0]
>> >  	return fmt.Sprintf("%s/%s (%s/%s) openshift/%s", path.Base(os.Args[0]), version, runtime.GOOS, runtime.GOARCH, commit)
>> >  }
>> > +
>> > +type HybridClient struct {
>>
>> See how the Kubernetes client is currently implemented:
>> https://github.com/kubernetes/kubernetes/blob/071d21257fdfd439a9286fffbd5972c85643cd49/pkg/client/unversioned/client.go#L128-l134
>>
>> We should do the same for our client
>>
>> This is what they do
>>
>> // Client is the implementation of a Kubernetes client.type Client struct {
>>     *RESTClient
>>     *ExtensionsClient
>>     // TODO: remove this when we re-structure pkg/client.
>>     *DiscoveryClient
>> }
>>
>> I don't think that this is a good thing. Having a single interface like
>> that encourages non-specificity and it does not scale well as you add
>> different clients. I think that a reasonable division is on the APIGroup
>> boundary. Using different clients for different APIGroups will help keep
>> the dependency structure clear, prevents a mega-client problem, and allows
>> a clear expectation as APIGroups are enabled and disabled.
>>
>> @smarterclayton <https://github.com/smarterclayton> as I recall, you
>> were not a fan combined Clients.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/5310/files#r42779351>.
>>
>
>

liggitt commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -293,6 +293,10 @@ type EtcdStorageConfig struct {
+	// KubernetesExpStorageVersion is the API version that Kube experimental resources in etcd
+	// should be serialized to. This value should *not* be advanced until all clients in the
+	// cluster that read from etcd have code that allows them to read the new version.
+	KubernetesExpStorageVersion string
```
@deads2k thoughts on this? this intersect or conflict with your work?

DirectXMan12 replied:
My plan was to just rebase on top of that work once it lands

liggitt commented on pkg/cmd/server/api/types.go:
```diff
@@ -332,6 +338,12 @@ type EtcdStorageConfig struct {
+	// KubernetesExpStorageVersion is the API version that Kube experimental resources in etcd
+	// should be serialized to. This value should *not* be advanced until all clients in the
+	// cluster that read from etcd have code that allows them to read the new version.
+	KubernetesExpStorageVersion string
```
don't call this experimental

liggitt commented on pkg/cmd/server/kubernetes/master.go:
```diff
@@ -106,6 +110,12 @@ func (c *MasterConfig) RunEndpointController() {
 
+func (c *MasterConfig) RunHorizontalPodAutoscalerController(oc *osclient.Client, kc *client.Client) {
+	hybridClient := osclient.NewHybridClient(oc, kc)
+	podautoscaler := podautoscalercontroller.NewHorizontalController(kc, hybridClient, kc, metrics.NewHeapsterMetricsClient(c.KubeClient))
```
use `kc`, not `c.KubeClient`

liggitt commented on pkg/cmd/server/origin/master.go:
```diff
@@ -436,7 +436,8 @@ func (c *MasterConfig) GetRestStorage() map[string]rest.Storage {
 
-		"deploymentConfigs":         deployConfigStorage,
+		"deploymentConfigs":         deployConfigStorage.DeploymentConfig,
+		"deploymentConfigs/scale":   deployConfigStorage.Scale,
```
need to update policy with any new resources (either resource group definitions in `authorization/api/types.go` or bootstrap roles in `cmd/server/bootstrappolicy/policy.go` or both)

kargakis commented:
Can you rebase on top of the latest master?

kargakis commented on pkg/deploy/api/v1/register.go:
```diff
@@ -9,9 +9,11 @@ func init() {
 func (*DeploymentConfig) IsAnAPIObject()         {}
 func (*DeploymentConfigList) IsAnAPIObject()     {}
 func (*DeploymentConfigRollback) IsAnAPIObject() {}
+func (*Scale) IsAnAPIObject() {}
```
Since this is going to be a dc subresource I would got with `DeploymentConfigScale`.

ncdc replied:
Upstream didn't do that. They just call it scale, and it's a subresource under both ReplicationController and Deployment.

deads2k commented on pkg/client/scale.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+// Get takes the reference to scale subresource and returns the subresource or error, if one occurs.
+func (c *scales) Get(kind string, name string) (result *extensions.Scale, err error) {
```
I don't think this abstraction is correct for our `client` API.  On the command line, yes: "please scale this resource for me".  In the code, no.  I'd like to see a scale operation added to the deploymentconfig client so that the compiler can properly enforce what makes sense and what doesn't.

ncdc replied:
This is required to work with the upstream HPA controller.

ncdc replied:
He's implementing https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/scale.go#L29, which is used by https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L99

deads2k replied:
> He's implementing https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/scale.go#L29, which is used by https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L99

```go
type delegatingScaler struct{
    dcClient DeploymentConfigInterface
    kubeScaler kclient.Scaler
}

func (s *delegatingScaler) Get(kind string, name string) (*extensions.Scale, error){
    switch (kind){
        "DeploymentConfig":
            return s.dcClient.GetScale(name)
    }

    return s.kubeScaler.Get(kind, name)
}
```

That keeps the list explicit, ensures that the client libraries provide solid types, prevents URL building against resources that aren't ours (belong to kube), allows proper error messages (instead of 403's or 404's) for origin types, and with upstream changes would allow a better location/delegation model.

liggitt replied:
+1

ncdc replied:
Thanks @deads2k !

DirectXMan12 commented:
I rebased and made the changes WRT the scale client.  The change to have it affect the current deployment should be up shortly

DirectXMan12 commented:
Ok, scale client affects current deployment now.

ncdc commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -1,26 +1,52 @@
+	Scale            *ScaleREST
+}
+
+func NewStorage(s storage.Interface, rcNamespacer kclient.ReplicationControllersNamespacer) DeploymentConfigStorage {
```
Godoc

ncdc commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +72,139 @@ func NewStorage(s storage.Interface) *REST {
 	return &REST{store}
 }
+
+type ScaleREST struct {
```
Godoc

DirectXMan12 commented:
@deads2k @liggitt this look good for the new way of accessing the scale subresources?

DirectXMan12 commented:
It also occurs to me that currently, the HPA expects heapster to be at "kube-system/heapster".  Where should/does it live for OpenShift?

liggitt commented:
That's hard coded? I would expect it to be configurable.

DirectXMan12 commented:
@liggitt behold: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/podautoscaler/metrics/metrics_client.go#L37

smarterclayton commented:
It's ok to upstream patch this first in this PR and get the follow up in
later.

On Oct 22, 2015, at 8:55 PM, Solly <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> behold:
https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/controller/podautoscaler/metrics/metrics_client.go#L37

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5310#issuecomment-150397529>.

deads2k commented on pkg/cmd/server/kubernetes/master.go:
```diff
@@ -11,6 +11,7 @@ import (
 
 	kapi "k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/record"
+	osclient "github.com/openshift/origin/pkg/client"
```
This dependency looks weird.  I think you need two controllers, one for origin and one for kube.  Otherwise, HPA won't work for DCs in the remote kube case.

DirectXMan12 replied:
So each controller would have to loop through all HPAs, run the check to see if the given scale resource pointed at Origin or Kube, and then skip it if it pointed at the wrong resource type?  That seems a bit redundant...

liggitt replied:
@deads2k what would a remote kube HPA controller do if it encountered a non kube resource reference?

deads2k replied:
> This dependency looks weird. I think you need two controllers, one for origin and one for kube. Otherwise, HPA won't work for DCs in the remote kube case.

Actually, once you split that, do you still need the union scaler?

ironcladlou replied:
If the controller were using `cache.Store` (& friends) and/or the controller/informer framework, couldn't each controller be configured during initialization to populate the stores with scale objects based on a field selector to get only kube/origin references?

ironcladlou replied:
Related: is there some reason the new controller isn't using the controller framework?

deads2k replied:
> @deads2k what would a remote kube HPA controller do if it encountered a non kube resource reference?

I don't see how it could do anything but fail.  There's no way to give it proper credentials and the current upstream impl will try to hit itself at `api/namespace/foo/deploymentconfigs/name/scale`, rigth?

deads2k replied:
> So each controller would have to loop through all HPAs, run the check to see if the given scale resource pointed at Origin or Kube, and then skip it if it pointed at the wrong resource type? That seems a bit redundant...

There's no other correct way to handle it.  A given HPA controller should only attempt to handle scale requests for kinds that it can understand.  It gets back to my points about being able to recognize which resources a scaler client is good for.

deads2k replied:
> This dependency looks weird. I think you need two controllers, one for origin and one for kube. Otherwise, HPA won't work for DCs in the remote kube case.

@smarterclayton Confirmed that HPA can be broken for external kubernetes cases for now.  I'll open an issue upstream to discuss how to fix this.

ironcladlou commented:
Scaling the deploymentConfig won't have any impact on rolling updates (which use the replica count from the old RC), and scaling an RC managed by a deploymentConfig directly will probably not be effective if a rolling update is in-progress (as the updater doesn't recompute its scaling inputs and doesn't use a reconciliation model like the new experimental deployment system.)

See also: #4327

I think there's probably more work to be done if this is to integrate cleanly with the current deployment system.

ironcladlou commented:
cc @derekwaynecarr @pmorie

DirectXMan12 commented:
@ironcladlou we don't actually scale the deploymentconfig in the deploymentconfig's scale subresource -- we scale the most recent deployment (assuming there is one, otherwise we just scale the config).  The scale subresource currently works more or less like the `kubectl scale dc` command.

I agree there are issues when in the middle of the rolling deployment though...  We may have to somehow say "don't scale while a deployment is deploying"

deads2k commented on pkg/client/deploymentconfigs.go:
```diff
@@ -23,6 +24,8 @@ type DeploymentConfigInterface interface {
 	Watch(label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
 	Generate(name string) (*deployapi.DeploymentConfig, error)
 	Rollback(config *deployapi.DeploymentConfigRollback) (*deployapi.DeploymentConfig, error)
+	GetScale(name string) (*extensions.Scale, error)
```
@ironcladlou would you prefer a `DeploymentConfigScalerInterface`?  I don't feel strongly.

deads2k replied:
@smarterclayton I think this will be the first cross group reference we have.  Has someone already considered the repercussions upstream?

deads2k commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -434,6 +434,11 @@ func (c *MasterConfig) ImageImportControllerClient() *osclient.Client {
 }
 
+// DeploymentConfigScaleClient returns the client used by the Scale subresource registry
+func (c *MasterConfig) DeploymentConfigScaleClient() *kclient.Client {
```
See the stub here: https://github.com/openshift/origin/pull/5357 for getting a properly scoped client.

DirectXMan12 replied:
This is for use inside the API server (it's used by the registry code of DeploymentConfigs to fetch the corresponding ReplicationControllers).  Does it still need to have a service account for that?

deads2k replied:
> This is for use inside the API server (it's used by the registry code of DeploymentConfigs to fetch the corresponding ReplicationControllers). Does it still need to have a service account for that?

Not yet.  This would be ok for that.

deads2k commented on pkg/deploy/api/v1/types.go:
```diff
@@ -370,3 +370,31 @@ type DeploymentLogOptions struct {
 }
+
+// ScaleSpec describes the attributes a Scale subresource
+type ScaleSpec struct {
```
There was weirdness and bad error messages.  Can you try https://github.com/openshift/origin/pull/5364 and see if you still need this?  I'd prefer to avoid having different scale objects if we can avoid it.  It might not be possible because deploymentconfigs are under `oapi/v1`, but I'd like to attempt it.

DirectXMan12 replied:
I tried, still need to have the Scale object in oapi/v1

kargakis commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +74,140 @@ func NewStorage(s storage.Interface) *REST {
+
+	// TODO(directxman12): this is going to be a bit out of sync, since we are calculating it
+	// here and not as part of the deploymentconfig loop -- is there a better way of doing it?
+	totalReplicas, err := r.replicasForDeploymentConfig(deployment.Namespace, deployment.Name)
```
Why are you summing all the replicas from all the deployment specs?

DirectXMan12 replied:
That seemed to be what upstream deployment did: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L246

kargakis commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +74,140 @@ func NewStorage(s storage.Interface) *REST {
+	// has been made
+	controller, err := r.rcNamespacer.ReplicationControllers(deployment.Namespace).Get(util.LatestDeploymentNameForConfig(deployment))
+	if err != nil {
+		if kerrors.IsNotFound(err) {
```
Please try to avoid nesting and `else` clauses. This can be boiled down to:
```go
if err != nil {
    if !kerrors.IsNotFound(err) {
        return nil, false, err
    }
    useDC = true
}
```

kargakis commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +74,140 @@ func NewStorage(s storage.Interface) *REST {
+	return &extensions.Scale{}
+}
+
+func (r *ScaleREST) Get(ctx kapi.Context, name string) (runtime.Object, error) {
```
godoc

kargakis commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +74,140 @@ func NewStorage(s storage.Interface) *REST {
+	}, nil
+}
+
+func (r *ScaleREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
```
godoc

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +74,140 @@ func NewStorage(s storage.Interface) *REST {
+	} else {
+		oldReplicas := controller.Spec.Replicas
+		controller.Spec.Replicas = scale.Spec.Replicas
+		if _, err = r.rcNamespacer.ReplicationControllers(deployment.Namespace).Update(controller); err != nil {
```
How are you making sure you aren't fighting with the deployment controller that's trying to scale this down?

DirectXMan12 replied:
This *should* be the latest deployment, so it shouldn't be being scaled down.  That being said, there is potentially a race with the controller here.  Perhaps we want to disallow scaling during a changeover?.  This is one of the issues that @ironcladlou mentioned.  If we did things more like upstream (where the deployment was reconciled with the template), this would be less of a problem here, since the controller manages everything.

deads2k replied:
> This should be the latest deployment, so it shouldn't be being scaled down. That being said, there is potentially a race with the controller here. Perhaps we want to disallow scaling during a changeover?. This is one of the issues that @ironcladlou mentioned. If we did things more like upstream (where the deployment was reconciled with the template), this would be less of a problem here, since the controller manages everything.

If you're in a state where you're creating a new deployment, this is either the RC N which should be brought up according to the unavailable and surge values or its the RC N-1 which should be brought down via the controller.  In either case, this code is wrong isn't it?  It would start fighting with the deployment controller.

ironcladlou replied:
For the benefit of other readers: this is resolved by checking the latest RC to ensure its phase is `Complete`.

DirectXMan12 commented:
@deads2k I updated the the Scale subresource to fail to update if an deployments are currently in any state besides "Complete" or "Failed" (this means that the HPA controller will error out on scaling any DCs currently doing and update). 

I also updated the HPA metrics client to take a configurable namespace and service name, and made it look in $OPENSHIFT_INFRA_NS/heapster instead of kube-system/heapster using that patch.

ironcladlou commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +73,166 @@ func NewStorage(s storage.Interface) *REST {
+	}
+
+	// check to make sure we don't have any deployments that are in progress
+	if inflight, err := r.inflightDeploymentsExist(deployment.Namespace, deployment.Name); err != nil{
```
Later on you look up the RC to target. What if you just check that single RC's status to ensure it's `Complete`? I don't think you'd need `inflightDeploymentsExist`.

DirectXMan12 replied:
Ok, updated.

ironcladlou commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +73,157 @@ func NewStorage(s storage.Interface) *REST {
+		useDC = true
+	}
+
+	if isDeploymentInflight(controller) {
```
I think this should be just:

```go
if util.DeploymentStatusFor(controller) != api.DeploymentStatusComplete {
  ...
}
```

Scaling up a failed deployment is probably not what we want to happen.

DirectXMan12 replied:
Good point.  Fixed.

kargakis commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +73,157 @@ func NewStorage(s storage.Interface) *REST {
+	return replicas, nil
+}
+
+func isDeploymentInflight(deployment *kapi.ReplicationController) bool {
```
There is already an utility for this: https://github.com/openshift/origin/blob/46d8271a97de24c07464198d28694eb4dd0072d7/pkg/deploy/util/util.go#L261

ironcladlou replied:
See https://github.com/openshift/origin/pull/5310/files#r43170677; we shouldn't autoscale a failed deployment.

deads2k commented on pkg/client/scale.go:
```diff
@@ -0,0 +1,52 @@
+	switch kind {
+	case "DeploymentConfig":
+		return c.dcs.GetScale(name)
+	default:
```
Check if its an origin kind to produce an error and better message here.

deads2k commented on pkg/client/scale.go:
```diff
@@ -0,0 +1,52 @@
+	switch kind {
+	case "DeploymentConfig":
+		return c.dcs.UpdateScale(scale)
+	default:
```
check for origin kind here too.

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -41,6 +41,11 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 				{
+					Verbs:     sets.NewString(authorizationapi.VerbAll),
+					Resources: sets.NewString(authorizationapi.ResourceAll),
+					APIGroups: []string{authorizationapi.APIGroupAll},
```
ought to be able to add the `APIGroups` bit to the already existing rule.

DirectXMan12 replied:
ACK.  Wasn't sure if "no api group" counted as something under "all API groups"

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -396,7 +401,42 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
+				},
+				{
+					Verbs:         sets.NewString("proxy"),
+					Resources:     sets.NewString("services/api"),
```
I'm not familiar with this subresource.  What does it do?

DirectXMan12 replied:
Effectively, this says "can proxy to the heapster service" -- this is what OpenShift complained about when the metrics client tries to proxy to heapster's API.

liggitt replied:
why are you going through the proxy to get to a service? why not go straight to the service?

liggitt replied:
this looks like a misparse of a proxy URL

liggitt replied:
https://github.com/kubernetes/kubernetes/issues/15830 is open to add `namespaces/ns/services/s/proxy`, but that doesn't exist yet

liggitt replied:
`services/api` is a bug... should be `services`... I'll open a fix upstream and send you a commit to cherry pick

liggitt replied:
add a big TODO here to fix the client to no longer need root proxy access

liggitt replied:
Opened https://github.com/kubernetes/kubernetes/pull/16570 to fix the URL parsing so you won't need to specify the `/api` subresource... please cherry pick that commit into this PR and fix the resource in this policy

deads2k commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -396,7 +401,42 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
+					Verbs:     sets.NewString("create", "update", "patch"),
+					Resources: sets.NewString("events"),
+				},
+				// Heapster MetricsClient
```
Add a //TODO to figure out how to scope this only to the heapster namespace.  This bit is kind of ugly.

@liggitt  Think we should we just require it in "infra" and create the rule there.  Someone could always add it to a different namespace later

deads2k replied:
Almost seems like this is a separate role.  Which of these is needed to talk to heapster?  I wasn't expecting "list pods" for heapster metric access.

DirectXMan12 replied:
"list pods" is for getting all pods for an RC, which we then pass to heapster to get the relevant metrics.

DirectXMan12 replied:
@deads2k if it's in infra, we have to create a new function here, since the actual name of infra is configurable, correct?

deads2k commented:
The more I think about the serialization from (origin, v1, scale) to (extensions, internal, scale) and back again, the more it feels like a bug.  Our scheme can do the conversion, but only because all internal kinds are sharing a namespace instead of being partitioned by group.  I don't think that most people would expect cross-group conversions to work and they only do work because group/version is considered opaque by the scheme which is shared across all groups.

The shared internal namespace of the scheme is a quirk that should be fixed, since its valid to have colliding kinds across groups.

deads2k commented:
Having our own serialization for `extensions, Scale` will produce a `deploymentconfigs/foo/scale` API that will never work with an external controller.

See https://github.com/kubernetes/kubernetes/issues/16177#issuecomment-151831609

DirectXMan12 commented:
I've integrated @deads2k's changes that allow dc/scale to accept/return extensions/v1beta1 Scale objects.

deads2k commented on pkg/client/deploymentconfigs.go:
```diff
@@ -107,3 +110,17 @@ func (c *deploymentConfigs) Rollback(config *deployapi.DeploymentConfigRollback)
+// Update updates an existing deploymentConfig
+func (c *deploymentConfigs) UpdateScale(scale *extensions.Scale) (result *extensions.Scale, err error) {
+	result = &extensions.Scale{}
+	err = c.r.Put().Namespace(c.ns).Resource("deploymentConfigs").Name(scale.Name).SubResource("scale").Body(scale).Do().Into(result)
```
Does this attempt to encode with the correct codec?  I'd be concerned about the client not knowing that it needed different codecs.

DirectXMan12 replied:
Well, it appears it currently attempts with the Origin v1 Codec, which won't work for us...

deads2k commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/apiserver/api_installer.go:
```diff
@@ -126,13 +126,21 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 	versionedObject := indirectArbitraryPointer(versionedPtr)
 
-	mapping, err := a.group.Mapper.RESTMapping(kind, a.group.Version)
+	// we need the correct version to serialize ehre
```
typo, here

deads2k replied:
> typo, here

ha, that was me.

deads2k commented on pkg/cmd/server/start/start_master.go:
```diff
@@ -547,7 +547,7 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		kc.RunReplicationController(rcClient)
 		kc.RunJobController(jobClient)
-		kc.RunHPAController(hpaKClient)
+		kc.RunHPAController(hpaOClient, hpaKClient, oc.Options.PolicyConfig.OpenShiftInfrastructureNamespace)
```
@liggitt Like this (and I think this is a good default) or should we have a separate knob?  I'd propose that as a followup if we want one.

DirectXMan12 commented:
Thanks to @deads2k, things should be working now with not having Scale in `oapi/v1`

liggitt commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -396,7 +397,42 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
+				},
+				// Heapster MetricsClient
+				{
+					Verbs:     sets.NewString("list"),
```
add a TODO to fix the metrics client to take pod names, rather than requiring cluster-wide pod-listing privileges

deads2k replied:
> add a TODO to fix the metrics client to take pod names, rather than requiring cluster-wide pod-listing privileges

Issue please.  That's big enough I'd like to make sure we don't lose it or wait until we come back to this code.

liggitt commented:
Can you cherry pick 15706 from upstream into this PR

deads2k commented:
@DirectXMan12 update the names of the UPSTREAM commits to indicate PR numbers if they have them.  Otherwise, indicate that they are carries or drops.

deads2k commented:
> Can you cherry pick 15706 from upstream into this PR

@liggitt do you really want it here?  I'd rather pull that in after this merges.

deads2k commented:
@DirectXMan12 I've added a TODO list for this PR to the description.  I think those things are all necessary before merge in addition to the rest of the review.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+}
+
+// Update scales the DeploymentConfig for the given Scale subresource, returning the updated Scale.
+func (r *ScaleREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object, bool, error) {
```
Add a call to validation like we discussed on IRC

deads2k replied:
Can you document the order in which the deploymentconfig controller does the updates to resources as it creates new RCs and changes its RC's replicas?  I need that information to reason about race potential.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+	if !ok {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("wrong object passed to Scale update: %v", obj))
+	}
+	deploymentConfig, err := (*r.registry).GetDeploymentConfig(ctx, scale.Name)
```
the dereference is unexpected.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+	}
+	deploymentConfig, err := (*r.registry).GetDeploymentConfig(ctx, scale.Name)
+	if err != nil {
+		return nil, false, errors.NewNotFound("scale", scale.Name)
```
only return `NotFound` if its actually `NotFound`, there's an `IsNotFound`

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+func (r *ScaleREST) Get(ctx kapi.Context, name string) (runtime.Object, error) {
+	deploymentConfig, err := (*r.registry).GetDeploymentConfig(ctx, name)
+	if err != nil {
+		return nil, errors.NewNotFound("scale", name)
```
check to see if this is actually not found

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+		useDC = true
+	}
+
+	if deploymentStatus := util.DeploymentStatusFor(controller); deploymentStatus != api.DeploymentStatusComplete {
```
`controller` can be nil here.  If it's nil, shouldn't we skip this check?

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+
+	// TODO(directxman12): this is going to be a bit out of sync, since we are calculating it
+	// here and not as part of the deploymentconfig loop -- is there a better way of doing it?
+	totalReplicas, err := r.replicasForDeploymentConfig(deploymentConfig.Namespace, deploymentConfig.Name)
```
Again, this doesn't make sense when `useDC` is true, right?  

It seems like the case of `useDC` should have early branch and return.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+
+	if useDC {
+		deploymentConfig.Template.ControllerTemplate.Replicas = scale.Spec.Replicas
+		if err = (*r.registry).UpdateDeploymentConfig(ctx, deploymentConfig); err != nil {
```
is the dereference needed?

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+
+// ScaleREST contains the REST storage for the Scale subresource of DeploymentConfigs.
+type ScaleREST struct {
+	registry     *deployconfig.Registry
```
a pointer to an interface is usually wrong.  Do you really need or mean it here?  It seems to have led to unexpected dereferences.

DirectXMan12 replied:
I don't think so, I think that's an artifact from copying from the Kubernetes code, which had it for some reason.  I'll take a look.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+		oldReplicas := controller.Spec.Replicas
+		controller.Spec.Replicas = scale.Spec.Replicas
+		if _, err = r.rcNamespacer.ReplicationControllers(deploymentConfig.Namespace).Update(controller); err != nil {
+			return nil, false, errors.NewConflict("scale", scale.Name, err)
```
Don't assume that this is a conflict error.  I think I'd just return the raw `err` here.

DirectXMan12 replied:
I concur.  I think that was also a artifact from the Kube code.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +75,150 @@ func NewStorage(s storage.Interface) *REST {
+	if useDC {
+		deploymentConfig.Template.ControllerTemplate.Replicas = scale.Spec.Replicas
+		if err = (*r.registry).UpdateDeploymentConfig(ctx, deploymentConfig); err != nil {
+			return nil, false, errors.NewConflict("scale", scale.Name, err)
```
don't return error types that aren't guaranteed accurate.

liggitt commented:
>> Can you cherry pick 15706 from upstream into this PR
>
> @liggitt do you really want it here? I'd rather pull that in after this merges.

I do... it looks like several similar files were touched and our commits will layer better long term if we build on what's already in master upstream

DirectXMan12 commented:
@deads2k do I need to add default policy for the Heapster pod itself, or is that just something that should be covered by docs?

deads2k commented on pkg/deploy/api/validation/validation.go:
```diff
@@ -47,6 +48,53 @@ func ValidateDeploymentConfigUpdate(newConfig *deployapi.DeploymentConfig, oldCo
 	return allErrs
 }
 
+func ValidateScale(scale *extensions.Scale) fielderrors.ValidationErrorList {
```
Why aren't we using upstream validators?

DirectXMan12 replied:
AFAICT, there *isnt* a validateScale upstream (I looked around)

deads2k replied:
> AFAICT, there isn't a validateScale upstream (I looked around)

Create one, make the upstream pull, pick down into here.

deads2k commented on pkg/deploy/api/validation/validation.go:
```diff
@@ -47,6 +48,53 @@ func ValidateDeploymentConfigUpdate(newConfig *deployapi.DeploymentConfig, oldCo
+	return allErrs
+}
+
+func ValidateScaleUpdate(newScale *extensions.Scale, oldScale *extensions.Scale) fielderrors.ValidationErrorList {
```
I would expect this to use upstream validation and we'd have one addition check for the kind.

deads2k commented:
I'm really hoping this is just my ignorance, but it looks like and HPA object (https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apis/extensions/types.go#L130) in one namespace can request via escalation scaling in another namespace: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/apis/extensions/types.go#L76.  

I would have expected a `LocalObjectReference`.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -10,17 +16,41 @@ import (
+// and their Scale subresources.
+type DeploymentConfigStorage struct {
+	DeploymentConfig *REST
+	Scale            *ScaleREST
```
This looks cyclical.  Is it dead?

DirectXMan12 replied:
??

deads2k replied:
> ??

Oh, yuck.  Another anti-pattern from upstream?  Different things should be different.  We can deal with it later.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+			Replicas: specReplicas,
+		},
+		Status: extensions.ScaleStatus{
+			Replicas: totalReplicas,
```
So when I have no RCs and my HPA is set to 6 replicas, wouldn't this make me return "I want 6 replicas, but I have 0 replicas" and that would cause the HPA controller to try to scale me to 6 over and over?

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+		specReplicas = controller.Spec.Replicas
+	}
+
+	return &extensions.Scale{
```
It seems like this can built ahead of time, use the if blocks to set the fields you care about to make it a little more clear.

```go
// initialize return Scale
controller, err := big line
if err != nil{
    if !erro....{
    }

    ret.SpecReplicas = 
    ret.TotalReplicas = 
    return ret, nil
}

ret.specReplicas = 
ret.TotalReplicas = 
return ret, nil
```

DirectXMan12 replied:
ACK

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("wrong object passed to Scale update: %v", obj))
+	}
+
+	deploymentConfig, err := r.registry.GetDeploymentConfig(ctx, scale.Name)
```
There should be a validate call in this flow somewhere.

deads2k replied:
Why not pull the existing using the `.Get` from this type?

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+		},
+	}
+
+	if errs := validation.ValidateScaleUpdate(scale, existing); len(errs) > 0 {
```
Ah, I see.  Move this above the point where you start using the Scale input.

DirectXMan12 replied:
ACK

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+		deploymentConfig.Template.ControllerTemplate.Replicas = scale.Spec.Replicas
+		if err = r.registry.UpdateDeploymentConfig(ctx, deploymentConfig); err != nil {
+			return nil, false, err
+		}
```
build the return object and return right here.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,183 @@ func NewStorage(s storage.Interface) *REST {
+		if err = r.registry.UpdateDeploymentConfig(ctx, deploymentConfig); err != nil {
+			return nil, false, err
+		}
+	} else {
```
This else should become unnecessary.

deads2k commented:
@smarterclayton @liggitt @ncdc @pweil- 

This still has some unexpected behavior with gets on DC's with no deployments and some races in the update path, but it gives us most of the functionality and I think the external API's shape is correct.  If we want this now, this is about as good as it gets.

Can I get a second review?

ironcladlou commented:
>This still has some unexpected behavior with gets on DC's with no deployments and some races in the update path,

Can you elaborate on what races you believe still exist?

deads2k commented:
> Can you elaborate on what races you believe still exist?

First one that comes to mind?  I have to master controllers running.  Both grab the current DC and then get the most recent RC, but  the copies are different.  Once sees RC1, the other sees RC2.  They both attempt to scale and now you have an old deployment running.

smarterclayton commented:
The next loop would observe the old one and scale it down?

deads2k commented:
> Can you elaborate on what races you believe still exist?

Second one that occurs to me is that I decide to update the DC, but a new deployerPod was already created.  The deployment controller stalls.  The deploypod is created, started, and gets the DC before we scale it.  We the HPA controller updates the DC.  The deployment controller unstalls and gets an update conflict.

I'm assuming there's deconflicting logic in there somewhere that prevents a deployer pod get before the DC update, but I didn't see it in my quick survey.

deads2k commented:
> The next loop would observe the old one and scale it down?

No, it would scale down the new one.

ironcladlou commented:
>The next loop would observe the old one and scale it down?

I don't think it will today. The scale-down-failed behavior relies on observing the deployer pod (which can get reaped), and even if the deployer pod's still around, I think there's a bug in the reconcilation loop where it'll skip the scaledown behavior if the status is already failed (so if it scaled down to 0 once and updated status to failed, it won't ever scale down again).

ironcladlou commented:
When I say "it" I'm referring to the deployment controllers.

ironcladlou commented:
I'm not sure how we'll close that multi-master race without moving to the upstream deployment model (where DC becomes the sole scaling target). What we're wrestling with now is trying to update DC+RC as an atomic unit, which of course is not going to happen.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,187 @@ func NewStorage(s storage.Interface) *REST {
+
+	// TODO(directxman12): this is going to be a bit out of sync, since we are calculating it
+	// here and not as part of the deploymentconfig loop -- is there a better way of doing it?
+	totalReplicas, err := r.replicasForDeploymentConfig(deploymentConfig.Namespace, deploymentConfig.Name)
```
@smarterclayton This sums the replica count of all RCs owned by the DC for total.  @DirectXMan12 said that's how it's done upstream.

deads2k commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -46,3 +76,187 @@ func NewStorage(s storage.Interface) *REST {
+
+	oldReplicas := controller.Spec.Replicas
+	controller.Spec.Replicas = scale.Spec.Replicas
+	if _, err = r.rcNamespacer.ReplicationControllers(deploymentConfig.Namespace).Update(controller); err != nil {
```
Add TODO to handle scale down by scaling down old deployment RCs in preference to new ones.  If we merge like this, we should make an issue to fix it.

DirectXMan12 replied:
We shouldn't be scaling at all unless there's only one RC in business, correct?

smarterclayton replied:
Since this is going to race against the controller, you cannot get into a state where the new RC is being created but the old RC has been completely scaled down.

deads2k replied:
> We shouldn't be scaling at all unless there's only one RC in business, correct?

I'm reasonably certain that its possible for HPA read a DC on thread A, DCController starts a new RC and scales down the old one on thread B, HPA reads old RC on threadA, HPA scales old RC on threadA.

Once that happens, then the `Get` will indicate that you have too many replicas running, the `Update` is called again and you scale down the new RC

DirectXMan12 replied:
Ah, I see what you're saying

deads2k commented:
lgetm

smarterclayton commented:
[merge] - can you link the follow up issue for races here and set it to p1?

deads2k commented:
> [merge] - can you link the follow up issue for races here and set it to p1?

created https://github.com/openshift/origin/issues/5597

deads2k commented:
gofmt 

```
hack/verify-gofmt.sh
!!! gofmt needs to be run on the following files: 
./pkg/client/scale.go
./pkg/client/testclient/fake_deploymentconfigs.go
./pkg/cmd/server/kubernetes/master.go
./pkg/deploy/registry/deployconfig/etcd/etcd.go
Try running 'gofmt -s -d [path]'
```

liggitt commented:
please run all `hack/verify-*.sh` scripts to help the poor merge queue

liggitt commented:
[test]

deads2k commented:
[merge]

deads2k commented:
removed merge.  Is it actually too late?

deads2k commented:
got ok from @smarterclayton [merge]
